lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java

```diff
@@ -6,159 +6,237 @@
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import org.graalvm.nativeimage.ImageInfo;
 import org.slf4j.LoggerFactory;
 
-public final class Parser implements AutoCloseable {
-  private static void initializeLibraries() {
-    try {
-      System.loadLibrary("enso_parser");
-      return;
-    } catch (LinkageError err) {
-      // try harder to find the library
-    }
-    String os = System.getProperty("os.name");
-    String name;
-    if (os.startsWith("Mac")) {
-      name = "libenso_parser.dylib";
-    } else if (os.startsWith("Windows")) {
-      name = "enso_parser.dll";
-    } else {
-      name = "libenso_parser.so";
-    }
+public final class Parser {
+  private Parser() {}
+
+  public static String getWarningMessage(Warning warning) {
+    return getWarningTemplate(warning.getId());
+  }
+
+  public static long isIdentOrOperator(CharSequence input) {
+    return getWorker().isIdentOrOperator(input);
+  }
+
+  public static ByteBuffer parseInputLazy(CharSequence input) {
+    return getWorker().parseInputLazy(input);
+  }
 
-    var whereAmI = Parser.class.getProtectionDomain().getCodeSource().getLocation();
-    File root;
-    try {
-      root = new File(whereAmI.toURI()).getParentFile();
-    } catch (URISyntaxException ex) {
-      root = new File(".").getAbsoluteFile();
+  public static Tree parse(CharSequence input) {
+    return getWorker().parse(input);
+  }
+
+  public static UUID getUuid(long metadata, long nodeOffset, long nodeLength) {
+    long high = getUuidHigh(metadata, nodeOffset, nodeLength);
+    long low = getUuidLow(metadata, nodeOffset, nodeLength);
+    if (high == 0 && low == 0) {
+      // The native interface uses the Nil UUID value as a marker to indicate that no UUID was
+      // attached.
+      // The Nil UUID will never collide with a real UUID generated by any scheme.
+      return null;
     }
-    try {
-      var d = root;
-      File path = null;
-      while (d != null) {
-        path = new File(new File(d, "component"), name);
-        if (path.exists()) break;
-        d = d.getParentFile();
-      }
-      if (d == null || path == null) {
-        throw new LinkageError("Cannot find parser in " + root);
+    return new UUID(high, low);
+  }
+
+  /* Worker-thread state */
+
+  private static final FinalizationManager finalizationManager = new FinalizationManager();
+
+  private static Worker getWorker() {
+    var worker = threadWorker.get();
+    if (worker == null) {
+      finalizationManager.runPendingFinalizers();
```

Comment:
I am surprised the cleanup of `runPendingFinalizers()` is called only when a new `Worker` for a thread is created. I would assume `runPendingFinalizers()` should be called more frequently - whenever `getWorker()` is called.